src/tools/chip-cert/Cmd_PrintTLV.cpp

```diff
@@ -0,0 +1,251 @@
+        return false;
+    }
+
+    gInFileNameOrStr = argv[0];
+
+    return true;
+}
+
+enum class TLVFormat
+{
+    kTLVFormat_Unknown = 0,
+    kTLVFormat_Raw,
+    kTLVFormat_Hex,
+    kTLVFormat_Base64,
+};
+
+std::optional<std::vector<uint8_t>> ReadTLV(const char * fileNameOrStr)
+{
+    TLVFormat tlvFmt = TLVFormat::kTLVFormat_Unknown;
+    size_t tlvLen    = 0;
+    std::vector<uint8_t> tlvBuf;
+
+    // If fileNameOrStr is a file name
+    if (access(fileNameOrStr, R_OK) == 0)
+    {
+        uint32_t computedSize = 0;
+        VerifyOrReturnError(ReadFileIntoMem(fileNameOrStr, nullptr, computedSize), std::nullopt);
+
+        tlvLen = computedSize;
+        tlvBuf.resize(tlvLen);
+
+        VerifyOrReturnError(ReadFileIntoMem(fileNameOrStr, tlvBuf.data(), computedSize), std::nullopt);
+        tlvFmt = TLVFormat::kTLVFormat_Raw;
+    }
+    // Otherwise, treat fileNameOrStr as a pointer to the TLV string (in hex or base64 encoded format)
+    else
+    {
+        tlvLen = strlen(fileNameOrStr);
+        tlvBuf.resize(tlvLen);
+        memcpy(tlvBuf.data(), fileNameOrStr, tlvLen);
+
+        // Check hex first since hex characters are a full subset of base64, and it's unlikely a
+        // large base64 payload would only have the hex space in it.
+        if (IsHexString(tlvBuf.data(), tlvLen))
+        {
+            tlvFmt = TLVFormat::kTLVFormat_Hex;
+        }
+        else if (IsBase64String(reinterpret_cast<const char *>(tlvBuf.data()), tlvLen))
+        {
+            tlvFmt = TLVFormat::kTLVFormat_Base64;
+        }
+        else
+        {
+            fprintf(stderr, "Invalid input string: neither hex nor base64 encoded!\n");
+            return std::nullopt;
+        }
+    }
+
+    if (tlvFmt == TLVFormat::kTLVFormat_Hex)
+    {
+        size_t len = chip::Encoding::HexToBytes(Uint8::to_char(tlvBuf.data()), tlvLen, tlvBuf.data(), tlvLen);
+        VerifyOrReturnError(CanCastTo<uint32_t>(2 * len), std::nullopt);
+        VerifyOrReturnError(2 * len == tlvLen, std::nullopt);
+        tlvLen = len;
+    }
+    else if (tlvFmt == TLVFormat::kTLVFormat_Base64)
+    {
+        VerifyOrReturnError(CanCastTo<uint16_t>(tlvLen), std::nullopt);
+        uint16_t decodedLen = chip::Base64Decode(Uint8::to_char(tlvBuf.data()), static_cast<uint16_t>(tlvLen), tlvBuf.data());
+        if (decodedLen == UINT16_MAX)
+        {
+            fprintf(stderr, "Invalid Base64 input string\n");
+            return std::nullopt;
+        }
+        tlvLen = decodedLen;
+    }
+
+    tlvBuf.resize(tlvLen);
+
+    return tlvBuf;
+}
+
+void ENFORCE_FORMAT(1, 2) SimpleDumpWriter(const char * aFormat, ...)
+{
+    va_list args;
+
+    va_start(args, aFormat);
+
+    vfprintf(gOutFile, aFormat, args);
+
+    va_end(args);
+}
+
+bool PrintTLV(ByteSpan tlv)
+{
+    chip::TLV::TLVReader reader;
+
+    VerifyOrReturnError(!tlv.empty(), false);
+    VerifyOrReturnError(OpenFile(gOutFileName, gOutFile, true), false);
```

Comment:
There should be a CloseFile somewhere too ... could we RAII these open/close bits?

It is probably ok because files get closed at exist, however does not feel clean.